### PR TITLE
update vaapi

### PR DESCRIPTION
**Pull request by alingherghescu**
- added new settings: platform, bitrate scale factor, low power, loop filter level, loop filter sharpness, async depth
- implemented new (dynamic) settings adjustment (in js)
- added new parameters: b, low_power, loop_filter_level, loop_filter_sharpness, async_depth
- tvhva_context_check_profile() will change TVHVAContext->entrypoint from VAEntrypointEncSlice into VAEntrypointEncSliceLP if VAEntrypointEncSlice is not available for that CODEC (according to VAAPI info)

### CODE REVIEW

**Comment by Flole998** [comment]
You can mark a PR as draft if you're not done yet, it's an option on the right side. I've done it for this PR now, when it's ready for review and merge just mark it as ready.

File: src/webui/static/app/codec.js [hunk 1]
```diff
@@ -229,9 +229,439 @@ var codec_profile_forms = {
...
+        });
+        // on rc_mode change
+        rc_mode_field.on('spin', function(spinner) {
+            updateFilters(form);
+        });
+    },
+
+    'codec_profile_vaapi_hevc': function(form) {
+        function updateFilters(form) {
+            //Ext.MessageBox.alert('updateFilters', 'updateFilters');
```
Comment (Flole998) [inline src/webui/static/app/codec.js:344]: This should be removed

File: src/webui/static/app/codec.js [hunk 2]
```diff
@@ -229,9 +229,439 @@ var codec_profile_forms = {
...
+        });
+        // on rc_mode change
+        rc_mode_field.on('spin', function(spinner) {
+            updateFilters(form);
+        });
+    },
+
+    'codec_profile_vaapi_vp8': function(form) {
+        function updateFilters(form) {
+            //Ext.MessageBox.alert('updateFilters', 'updateFilters');
```
Comment (Flole998) [inline src/webui/static/app/codec.js:454]: Same here

File: src/webui/static/app/codec.js [hunk 3]
```diff
@@ -229,9 +229,439 @@ var codec_profile_forms = {
...
+        // on rc_mode change
+        rc_mode_field.on('spin', function(spinner) {
+            updateFilters(form);
+        });
+    },
 
-    'codec_profile_vaapi_hevc': genericCBRvsVBR
+    'codec_profile_vaapi_vp9': function(form) {
+        function updateFilters(form) {
+            //Ext.MessageBox.alert('updateFilters', 'updateFilters');
```
Comment (Flole998) [inline src/webui/static/app/codec.js:561]: And here

File: src/transcoding/codec/codecs/libs/vaapi.c [hunk 4]
```diff
@@ -280,24 +451,147 @@ static int
...
-        AV_DICT_SET_INT(opts, "bufsize", ((self->bit_rate) * 1000) * self->buff_factor, AV_DICT_DONT_OVERWRITE);
-        AV_DICT_SET(opts, "force_key_frames", "expr:gte(t,n_forced*3)", AV_DICT_DONT_OVERWRITE);
+    // to avoid issues we have this check:
+    if (self->buff_factor <= 0) {
+        self->buff_factor = 3;
+    }
+    int int_bitrate = (int)((double)(self->bit_rate) * 1024.0 * (1.0 + (self->bit_rate_scale_factor * ((double)(self->size.den) - 480.0) / 480.0)));
+    int int_buffer_size = (int)((double)(self->bit_rate) * 2048.0 * self->buff_factor * (1.0 + self->bit_rate_scale_factor * ((double)(self->size.den) - 480.0) / 480));
+    int int_max_bitrate = (int)((double)(self->max_bit_rate) * 1024.0 * (1.0 + (self->bit_rate_scale_factor * ((double)(self->size.den) - 480.0) / 480.0)));
+    //tvherror(LS_VAAPI, "bitrate = %d, max_bitrate = %d, buffer_size = %d", int_bitrate, int_max_bitrate, int_buffer_size);
```
Comment (Flole998) [inline src/transcoding/codec/codecs/libs/vaapi.c:461]: Same as above

File: src/transcoding/codec/codecs/libs/vaapi.c [hunk 5]
```diff
@@ -347,22 +652,165 @@ static int
...
+    // to avoid issues we have this check:
+    if (self->buff_factor <= 0) {
+        self->buff_factor = 3;
     }
-    else {
-        AV_DICT_SET_QP(opts, self->qp, 25);
+    int int_bitrate = (int)((double)(self->bit_rate) * 1024.0 * (1.0 + (self->bit_rate_scale_factor * ((double)(self->size.den) - 480.0) / 480.0)));
+    int int_buffer_size = (int)((double)(self->bit_rate) * 2048.0 * self->buff_factor * (1.0 + self->bit_rate_scale_factor * ((double)(self->size.den) - 480.0) / 480));
+    int int_max_bitrate = (int)((double)(self->max_bit_rate) * 1024.0 * (1.0 + (self->bit_rate_scale_factor * ((double)(self->size.den) - 480.0) / 480.0)));
+    //tvherror(LS_VAAPI, "bitrate = %d, max_bitrate = %d, buffer_size = %d", int_bitrate, int_max_bitrate, int_buffer_size);
```
Comment (Flole998) [inline src/transcoding/codec/codecs/libs/vaapi.c:662]: Same as above

**Comment by alingherghescu** [comment]
I fixed all of the above ... but there is room for improvement; still working on this update.

**Comment by alingherghescu** [comment]
I pushed a new set of changes (and continue assuming we figure out the AMD issue) in order for you to check only the differences. After we finish, I will close this PR and perform a single submit.

**Comment by alingherghescu** [comment]
done. 
Let me know when you finished the review and I will close this PR and do a single submit. For sure some small comments are left hanging ...
I don't know how to merge the submits so I will close the PR and open a new one.
If you prefer to review all the changes at once (on the new submit) let me know and I can close this PR and open one clean PR with all the changes.

**Comment by alingherghescu** [comment]
I tried to enable the draft but I only see: Still in progress? Convert it to draft
Should I use this?
Gave me a warning when I press on that:
image

**Comment by Flole998** [comment]
That's to convert it back to a draft.

File: src/transcoding/codec/codecs/libs/vaapi.c [hunk 6]
```diff
@@ -22,6 +22,39 @@
...
+        { N_("Intel"),          1 },
+        { N_("AMD"),            2 },
+    };
+    return strtab2htsmsg(tab, 1, lang);
+}
+
+static htsmsg_t *
+rc_mode_get_list( void *o, const char *lang )
+{
+    // -1=skip 0=auto 1=CQP 2=CBR 3=VBR 4=ICQ 5=QVBR 6=AVBR
```
Comment (Flole998) [inline src/transcoding/codec/codecs/libs/vaapi.c:45]: No need to add this as a comment as the table is just below this.

File: src/transcoding/codec/codecs/libs/vaapi.c [hunk 7]
```diff
@@ -151,6 +183,17 @@ static const codec_profile_class_t codec_profile_vaapi_class = {
                 .off      = offsetof(tvh_codec_profile_vaapi_t, device),
                 .list     = tvh_codec_profile_vaapi_device_list,
             },
+            {
+                .type     = PT_BOOL,
+                .id       = "low_power",     // Don't change
+                .name     = N_("Low Power (low_power)"),
```
Comment (Flole998) [inline src/transcoding/codec/codecs/libs/vaapi.c:189]: Why is the name repeated in parentheses?

File: src/transcoding/transcode/hwaccels/vaapi.c [hunk 8]
```diff
@@ -256,30 +262,56 @@ tvhva_context_check_profile(TVHVAContext *self, VAProfile profile)
...
                 }
             }
-            if (res != 0) {
-                // if VAEntrypointEncSlice was not available we try again with VAEntrypointEncSliceLP
+        }
+        else {
+            tvherror(LS_VAAPI, "%s: va_res != VA_STATUS_SUCCESS; Failed to query entrypoints: %d (%s), run: $ vainfo", self->logpref, va_res, vaErrorStr(va_res));
+        }
+        if (res != 0) {
+            // before gaving up we swap VAEntrypointEncSliceLP with VAEntrypointEncSlice or viceversa
```
Comment (Flole998) [inline src/transcoding/transcode/hwaccels/vaapi.c:282]: *giving

File: src/transcoding/codec/codecs/libs/vaapi.c (hunk 7 above)
Comment (alingherghescu) [reply src/transcoding/codec/codecs/libs/vaapi.c:189]: the users that use ffmpeg are used with some args (low_power in this case) is the name of the parameter from ffmpeg.
maybe is foolish and we should remove it ... if you think are misleading.

File: src/transcoding/transcode/hwaccels/vaapi.c (hunk 8 above)
Comment (alingherghescu) [reply src/transcoding/transcode/hwaccels/vaapi.c:282]: done

File: src/transcoding/codec/codecs/libs/vaapi.c (hunk 7 above)
Comment (Flole998) [reply src/transcoding/codec/codecs/libs/vaapi.c:189]: Something like this can be mentioned in the docs, it shouldn't be part of the name of that option.

File: src/transcoding/codec/codecs/libs/vaapi.c (hunk 6 above)
Comment (alingherghescu) [reply src/transcoding/codec/codecs/libs/vaapi.c:45]: done

File: src/transcoding/codec/codecs/libs/vaapi.c (hunk 7 above)
Comment (alingherghescu) [reply src/transcoding/codec/codecs/libs/vaapi.c:189]: good point ... let me remove them.

Comment (alingherghescu) [reply src/transcoding/codec/codecs/libs/vaapi.c:189]: Done

**Comment by alingherghescu** [comment]
Do you want me to push here all the changes requested on this PR OR should I close this PR and open a clean one with all the changes in?

**Comment by Flole998** [comment]
You can also squash your changes and then do a force-push, then it's still here but as a single commit.

**Comment by alingherghescu** [comment]
the only way I know is: delete my branch tvheadend and start from scratch :(

**Comment by Flole998** [comment]
That's simple:

`git rebase -i HEAD~5` (as you have 5 commits), "pick" the first one, "squash" the following ones, give it a meaningful message that describes all changes, then `git push -f`.

**Comment by alingherghescu** [comment]
sorry .... me and github are not BFF :(

**Comment by alingherghescu** [comment]
https://github.com/tvheadend/tvheadend/pull/1490

**Comment by alingherghescu** [comment]
basically I got stuck to 'pick' part (didn't know how to pick) ... and closed the CMD. On the second open the result was different after I sent the command ... like I said me and github are not friends.

**Comment by Flole998** [comment]
The beginning of the line indicates what you want to do with that commit. It should be pick as default, so just leave it, change all the others to squash, done. Also most likely you created a new commit now, so increase the 5 by one and you can probably try again.

It's really not difficult and doesn't pollute the project with a dozen PRs that are all abandoned and superseded. Conversations are also hard to follow later on if they are spread across a bunch of closed PRs.

**Comment by jansohn** [comment]
@alingherghescu try these commands to squash the commits (please check the number of commits you want to squash):

```
git reset --soft HEAD~6
git commit -m "your commit message"
git push -f
```

**Comment by alingherghescu** [comment]
sorry, but I deleted all folder 'tvheadend' and restarted from scratch (clone + replace some files).
So I can't send the commands now :( .

**Comment by Saentist** [comment]
`git diff` to your local code ;)

**Comment by Flole998** [comment]
No, `git diff` doesn't help in that case. It would be necessary to set the upstream branch to the branch/repo this PR is based on and then force-push to it to recover from this now.

**Comment by Saentist** [comment]
@Flole998 
So many problems, because you don't ask for permission, 
to have access to squash function build-in in GitHub.

Who need to ask for you, to give this permissions to you?

**Comment by Flole998** [comment]
No because people don't know how to use git properly.

Not entirely sure why it's so hard for you to understand that it's not possible to squash-rebase-merge. That's simply not an option, and this is not the first time I'm explaining that to you. Please don't give wrong advice if you don't know what you're talking about.

**Comment by alingherghescu** [comment]
let's don't create more animosity in the community on this subject. The problem here is me (my inability to understand how github works). I know two commands: github clone and github pull. Everything else was was done with githubDesktop. I am not using github in my development at all.

**Comment by alingherghescu** [comment]
how I am performing development? I used Visual Studio Code on Windows 10. After I change a file I copy with winscp to Lubuntu (in my github clone folder). Build over SSH the source code and check the results. In order to share this with community I have to use githubdesstop to close on my Windows and replace the files from ~/Document/GitHub/tvheadend/ .... After that there is a learn/guess what I need to press to push that to tvheadend. Likely I don't have access so GithubDesktop ask me if I don't want to clone this project in my state space ... from here I move on the browser where there is another learn/guess of buttons to generate the PR :(

**Comment by Saentist** [comment]
> No because people don't know how to use git properly.
> 
> Not entirely sure why it's so hard for you to understand that it's not possible to squash-rebase-merge. That's simply not an option, and this is not the first time I'm explaining that to you. Please don't give wrong advice if you don't know what you're talking about.

https://egghead.io/lessons/github-automatically-squash-commits-on-github-when-merging-pull-requests

**Comment by Flole998** [comment]
> > No because people don't know how to use git properly.
> > Not entirely sure why it's so hard for you to understand that it's not possible to squash-rebase-merge. That's simply not an option, and this is not the first time I'm explaining that to you. Please don't give wrong advice if you don't know what you're talking about.
> 
> https://egghead.io/lessons/github-automatically-squash-commits-on-github-when-merging-pull-requests

That's for squash-merging, now show me how to do squash-rebase-merge. You are making the same wrong assumptions as last time and are wasting my time. Again: Please don't give wrong advice if you don't know what you're talking about.

**Comment by Saentist** [comment]
image
And what do this?

**Comment by Flole998** [comment]
It enables the option to rebase-merge. That's what we are doing in this project and if commits are not squashed we will end up with dozens of useless commits.

Why don't you create a repository, open pull requests against that repository and mess around with the options? Then you can learn what all features do and try to figure out a way to squash-rebase-merge as long as you want. Here that's offtopic.

**Comment by jansohn** [comment]
> It enables the option to rebase-merge. That's what we are doing in this project and if commits are not squashed we will end up with dozens of useless commits.

I'm curious and hope you don't mind explaining it one more time as I couldn't find the answer in the old PRs. This repo has "Allow squash merging" disabled and since you don't have admin access you cannot change it?

**Comment by Flole998** [comment]
It's simply not an option to squash-rebase-merge, no matter what you enable or disable. This project doesn't have many guidelines when it comes to contributing, so I really don't understand what the problem is. Other projects require commits to be signed-off or even have checks for formatting.

**Comment by alingherghescu** [comment]
I have a practical question: If I want to develop a new feature should I start from my tvheadend (with changes already in) or from master tvheadend (without my previous updates)?

**Comment by Flole998** [comment]
You always start from master, everything else will be difficult to merge.

**Comment by alingherghescu** [comment]
I plan to add a check-box: Resize only down (in Profile Settings: profile_video_class.c) --> in order to resize only resolutions higher than target; for example if you set 720 it will not resize 480 or 720 but will resize down 1080->720. This will affect also nvidia (I think). Are you OK will this change?

**Comment by Flole998** [comment]
Yeah sure, great idea! Please name it "Disable upscaling" that's probably an easier to understand name for it.

**Comment by Saentist** [comment]
> I plan to add a check-box: Resize only down (in Profile Settings: profile_video_class.c) --> in order to resize only resolutions higher than target; for example if you set 720 it will not resize 480 or 720 but will resize down 1080->720. This will affect also nvidia (I think). Are you OK will this change?

More logical is to be called Upscale Downscale 
check https://tvheadend.org/issues/3131

**Comment by jansohn** [comment]
> This project doesn't have many guidelines when it comes to contributing, so I really don't understand what the problem is. Other projects require commits to be signed-off or even have checks for formatting.

It would make life easier for (first time) contributors since many have very limited knowledge of Git. And since this project has no active developers it probably needs these contributions "to survive".

Squash merge has the benefit that it also links to the PR number. It produces only one commit on top of the history. IMO the only difference to manual squash and then rebase-merge is a slightly different commit message (depending on what you select during merge of course).

This is just friendly advice though so keep doing what you prefer. I'm happy there is still a person that has write access to this project. Did you contact contributors with admin access before to solve the missing admin rights problem?

**Comment by alingherghescu** [comment]
> More logical is to be called Upscale Downscale
> check https://tvheadend.org/issues/3131

We have to come to some agreement ... before I start implementing. We can have a combo with 3 options: Upscale/Downscale (what we have today), Upscale and Downscale --> but frankly why would you use only the Upscale? 
Regarding the feature request: #3131 my implementation is covering (roughly) that request. Maybe the way is implemented is not using SD/HD/UHD as decision. I also started from the some need (to encode different streams with different parameters), but I realized that based on location SD might be 480 or 576; also HD might be 720 or 1080.

> if its possible for SD aspect ratio check (4:3/16:9)
This part is missing; The algorithm is only using Height to adjust bitrate, maxbitrate and buffer.
--> if is required, we can replace Height with Height * Width (to account for number of pixels per image).

> Something similar with audio
Not covering this part.

**Comment by Flole998** [comment]
It all depends on what you want to implement: If you just want to disable upscaling then "disable upscaling" is the correct name. If you want to implement a combobox then something like "scaling mode" would be more appropriate.

**Comment by Flole998** [comment]
> > This project doesn't have many guidelines when it comes to contributing, so I really don't understand what the problem is. Other projects require commits to be signed-off or even have checks for formatting.
> 
> It would make life easier for (first time) contributors since many have very limited knowledge of Git. And since this project has no active developers it probably needs these contributions "to survive".

Doesn't really matter. Let's face it: Tvheadend is mostly dead and professional developers know how to squash (or even just amend), those would be needed. Such a software can't be maintained by people who don't know how to use git as they lack other knowledge then aswell.

> Squash merge has the benefit that it also links to the PR number. It produces only one commit on top of the history. IMO the only difference to manual squash and then rebase-merge is a slightly different commit message (depending on what you select during merge of course).

You don't see at the commit who merged it. That's the disadvantage for the other options.

> This is just friendly advice though so keep doing what you prefer. I'm happy there is still a person that has write access to this project. Did you contact contributors with admin access before to solve the missing admin rights problem?

I did, still I'm not set as owner for the Tvheadend organization here on GitHub.

**Comment by Saentist** [comment]
> > More logical is to be called Upscale Downscale
> > check https://tvheadend.org/issues/3131
 
> We have to come to some agreement ... before I start implementing. We can have a combo with 3 options: Upscale/Downscale (what we have today), Upscale and Downscale --> but frankly why would you use only the Upscale? Regarding the feature request: #3131 my implementation is covering (roughly) that request. Maybe the way is implemented is not using SD/HD/UHD as decision. I also started from the some need (to encode different streams with different parameters), but I realized that based on location SD might be 480 or 576; also HD might be 720 or 1080.

 
> > if its possible for SD aspect ratio check (4:3/16:9)

> This part is missing; The algorithm is only using Height to adjust bitrate, maxbitrate and buffer.
> --> if is required, we can replace Height with Height * Width (to account for number of pixels per image).

aspect ratio info 
https://tvheadend.org/issues/4599

**Comment by alingherghescu** [comment]
@Flole998 the "scaling mode" is ready. Let me know when you finish with #1490 ... because I don't want to mess up with update vaapi branch. If someone has nvidia more testing would help. Replace files: internals.h and profile_video_class.c in folder src/transcoding/codec with attached ones. The debug window will will print input resolution and output resolution (IN and OUT for this function).
[source_code.zip](https://github.com/tvheadend/tvheadend/files/10369649/source_code.zip)

**Comment by alingherghescu** [comment]
good news :) I manages to squash 3 commits with githubDesktop.
before_git
after_git

**Comment by alingherghescu** [comment]
> > This part is missing; The algorithm is only using Height to adjust bitrate, maxbitrate and buffer.
> > --> if is required, we can replace Height with Height * Width (to account for number of pixels per image).
> 
> aspect ratio info https://tvheadend.org/issues/4599

The way I understand the code, we keep the aspect ration. The SAR is kept the same between IN and OUT after resize. In my research I found this post : https://video.stackexchange.com/questions/20871/how-do-i-convert-anamorphic-hdv-video-to-normal-h-264-video-with-ffmpeg-how-to that is showing a different way to resize (using SAR); but I don't think is a problem in the way resize is done on TVH today. Let me know if you disagree or you have any info showing otherwise.

**Comment by alingherghescu** [comment]
@Flole998 Can I provide a PR with 2 different submits (with two different purposes)? If one is fixing a bug and one is adding a feature (for example) ... should I merge those commits?

**Comment by Flole998** [comment]
You can put it in 1 PR but in that case with 2 different commits, one for each change. It's probably easier though to just do 2 PRs.

**Comment by alingherghescu** [comment]
Let's go slow and steady, first PR is ready: https://github.com/tvheadend/tvheadend/pull/1491
